File: Java/BlackMarket.java
import java.util.List;

public class BlackMarket extends Market {
    private String secretCode;
    private String hiddenLocation;

    // Default constructor
    public BlackMarket() {
        super(); // panggil constructor default Market
        this.secretCode = "";
        this.hiddenLocation = "";
    }

    // Parameterized constructor
    public BlackMarket(int idMarket, String name, String location, String hours,
                       String secretCode, String hiddenLocation) {
        super(idMarket, name, location, hours); // panggil constructor parent
        this.secretCode = secretCode;
        this.hiddenLocation = hiddenLocation;
    }

    // Getter methods
    public String getSecretCode() {
        return secretCode;
    }

    public String getHiddenLocation() {
        return hiddenLocation;
    }

    // Setter methods
    public void setSecretCode(String code) {
        this.secretCode = code;
    }

    public void setHiddenLocation(String location) {
        this.hiddenLocation = location;
    }

    // Display all info hierarchically
    public void displayBlackMarket(int indent) {
        String sp = " ".repeat(indent);
        System.out.println(sp + "Secret Code     : " + getSecretCode());
        System.out.println(sp + "Hidden Location : " + getHiddenLocation());

        // Display Market attributes
        System.out.println(sp + "Market ID       : " + getIdMarket());
        System.out.println(sp + "Name            : " + getMarketName());
        System.out.println(sp + "Location        : " + getMarketLocation());
        System.out.println(sp + "Operating Hours : " + getOperatingHours());
        System.out.println();

        // Display Sellers and their products
        if (!getSellerList().isEmpty()) {
            System.out.println(sp + "    Sellers:");
            for (Seller seller : getSellerList()) {
                seller.displaySeller(indent + 8);
            }
        }
        System.out.println();
    }

    // Overload untuk default indent = 0
    public void displayBlackMarket() {
        displayBlackMarket(0);
    }
}

File: Cpp/BlackMarket.cpp
#include <iostream>
#include <string>
#include <vector>
#include "Market.cpp"

using namespace std;

class BlackMarket : public Market {
private:
    string SecretCode;
    string HiddenLocation;

public:
    // Default constructor
    BlackMarket() : Market() {
        SecretCode = "";
        HiddenLocation = "";
    }

    // Parameterized constructor
    BlackMarket(int idMarket, string name, string location, string hours,
                string secretCode, string hiddenLocation)
        : Market(idMarket, name, location, hours) {
        this->SecretCode = secretCode;
        this->HiddenLocation = hiddenLocation;
    }

    // Getter methods
    string getSecretCode() {
        return this->SecretCode;
    }

    string getHiddenLocation() {
        return this->HiddenLocation;
    }

    // Setter methods
    void setSecretCode(string code) {
        this->SecretCode = code;
    }

    void setHiddenLocation(string location) {
        this->HiddenLocation = location;
    }

    // Display all info hierarchically
    void displayBlackMarket(int indent = 0) {
        string sp(indent, ' ');
        cout << sp << "Secret Code     : " << getSecretCode() << endl;
        cout << sp << "Hidden Location : " << getHiddenLocation() << endl;

        // Display Market attributes
        cout << sp << "Market ID       : " << getIdMarket() << endl;
        cout << sp << "Name            : " << getMarketName() << endl;
        cout << sp << "Location        : " << getMarketLocation() << endl;
        cout << sp << "Operating Hours : " << getOperatingHours() << endl << endl;

        // Display Sellers and their products
        if(!getSellerList().empty()){
            cout << sp << "    Sellers:" << endl;
            for(auto &seller : getSellerList()){
                seller.displaySeller(indent + 8);
            }
        }
        cout << endl;
    }

    // Destructor
    ~BlackMarket() {
    }
};
